src/crimson/admin/admin_socket.h

```diff
@@ -0,0 +1,306 @@
+ */
+struct AsokServiceDef {
+  const std::string command;  ///< the sequence of words that should be used
+  const std::string cmddesc;  ///< the command syntax
+  const AdminSocketHook* hook;
+  const std::string help;  ///< help message
+};
+
+class AdminHooksIter;  ///< an iterator over all APIs in all server blocks
+
+/// a ref-count owner of the AdminSocket, used to guarantee its existence until
+/// all server-blocks are unregistered
+using AdminSocketRef = seastar::lw_shared_ptr<AdminSocket>;
+
+using AsokRegistrationRes =
+  std::optional<AdminSocketRef>;  // holding the server alive until after our
+                                  // unregistration
+
+class AdminSocket : public seastar::enable_lw_shared_from_this<AdminSocket> {
+ public:
+  AdminSocket() = default;
+  ~AdminSocket() = default;
+
+  AdminSocket(const AdminSocket&) = delete;
+  AdminSocket& operator=(const AdminSocket&) = delete;
+  AdminSocket(AdminSocket&&) = delete;
+  AdminSocket& operator=(AdminSocket&&) = delete;
+
+  using hook_server_tag = const void*;
+
+  /**
+   *  create the async Seastar thread that handles asok commands arriving
+   *  over the socket.
+   */
+  seastar::future<> init(const std::string& path);
+
+  seastar::future<> stop();
+
+  /**
+   * register an admin socket hooks server
+   *
+   * The server registers a set of APIs under a common hook_server_tag.
+   *
+   * Commands (APIs) are registered under a command string. Incoming
+   * commands are split by spaces and matched against the longest
+   * registered command. For example, if 'foo' and 'foo bar' are
+   * registered, and an incoming command is 'foo bar baz', it is
+   * matched with 'foo bar', while 'foo fud' will match 'foo'.
+   *
+   * The entire incoming command string is passed to the registered
+   * hook.
+   *
+   * \param server_tag  a tag identifying the server registering the hook
+   * \param apis_served a vector of the commands served by this server. Each
+   *        command registration includes its identifying command string, the
+   *        expected call syntax, and some help text.
+   *
+   * A note regarding the help text: if empty, command will not be
+   * included in 'help' output.
+   *
+   * \retval a shared ptr to the asok server itself, or nullopt if
+   *         a block with same tag is already registered.
+   */
+  seastar::future<AsokRegistrationRes> register_server(
+    hook_server_tag server_tag, const std::vector<AsokServiceDef>& apis_served);
+
+  /**
+   * unregister all hooks registered by this hooks-server. The caller
+   * gives up on its shared-ownership of the asok server once the
+   * deregistration is complete.
+   */
+  seastar::future<> unregister_server(hook_server_tag server_tag,
+                                      AdminSocketRef&& server_ref);
+
+ private:
+  /**
+   * the result of analyzing an incoming command, and locating it in
+   * the registered APIs collection.
+   */
+  struct parsed_command_t {
+    std::string m_cmd;
```

Comment:
no need to prefix member variables of plain data struct with "m_" -- there is no chance to use them without the enclosing type.

Reply:
OK